osu.Game/Screens/Play/GameplayClockContainer.cs

```diff
@@ -160,6 +160,13 @@ public void Reset(double? time = null, bool startClock = false)
 
             Seek(StartTime);
 
+            // This is a workaround for the fact that DecoupleableInterpolatingFramedClock doesn't seek the source
+            // if the source is not IsRunning. (see https://github.com/ppy/osu-framework/blob/2102638056dfcf85d21b4d85266d53b5dd018767/osu.Framework/Timing/DecoupleableInterpolatingFramedClock.cs#L209-L210)
+            //
+            // This breaks in multiplayer spectator.
+            // I hope to remove this once we knock some sense into clocks in general.
+            (SourceClock as IAdjustableClock)?.Seek(StartTime);
```

Comment:
So here's a "fun" question. This is preceded by a call to `Seek()`, which is:

https://github.com/ppy/osu/blob/9da1de84b5452213cddb991dcf21bac33b760dd7/osu.Game/Screens/Play/GameplayClockContainer.cs#L117-L128

Which in turn calls

https://github.com/ppy/osu/blob/9da1de84b5452213cddb991dcf21bac33b760dd7/osu.Game/Beatmaps/FramedBeatmapClock.cs#L186-L192

Uh oh. Now there's offsets and decoupled interpolations and so on. The offsets worry me especially. It appears that in standard gameplay this currently seems to be partially saved by (a) the fact that the clock initially stops, which gets rid of the interpolation, and (b) the fact that offsets are applied in `LoadComplete()` and so they're zero at the point where `Reset()` is currently called, which is:

https://github.com/ppy/osu/blob/9da1de84b5452213cddb991dcf21bac33b760dd7/osu.Game/Beatmaps/FramedBeatmapClock.cs#L101-L121

but in editor gameplay test I can get this log to trigger:

```diff
diff --git a/osu.Game/Screens/Play/GameplayClockContainer.cs b/osu.Game/Screens/Play/GameplayClockContainer.cs
index 226108209a..70d4e17110 100644
--- a/osu.Game/Screens/Play/GameplayClockContainer.cs
+++ b/osu.Game/Screens/Play/GameplayClockContainer.cs
@@ -167,6 +167,9 @@ public void Reset(double? time = null, bool startClock = false)
             // I hope to remove this once we knock some sense into clocks in general.
             (SourceClock as IAdjustableClock)?.Seek(StartTime);
 
+            if (SourceClock.CurrentTime != CurrentTime)
+                Logger.Log($"UH OH WE DID A DESYNC. gcc is {CurrentTime}, source is {SourceClock.CurrentTime}");
+
             if (!wasPaused || startClock)
                 Start();
         }

```

Am I thinking correctly that this is a potential desync? And if so, then _maybe_ just for sheer safety this should be

```diff
diff --git a/osu.Game/Screens/Play/GameplayClockContainer.cs b/osu.Game/Screens/Play/GameplayClockContainer.cs
index 226108209a..35e1e0b792 100644
--- a/osu.Game/Screens/Play/GameplayClockContainer.cs
+++ b/osu.Game/Screens/Play/GameplayClockContainer.cs
@@ -165,7 +165,7 @@ public void Reset(double? time = null, bool startClock = false)
             //
             // This breaks in multiplayer spectator.
             // I hope to remove this once we knock some sense into clocks in general.
-            (SourceClock as IAdjustableClock)?.Seek(StartTime);
+            (SourceClock as IAdjustableClock)?.Seek(CurrentTime);
 
             if (!wasPaused || startClock)
                 Start();

```

Reply:
I think this makes sense. I've updated the workaround and comment to hopefully explain this requirement.